Add text for RFC76 OGR Python drivers (#1984)
OSGeo/gdal: gdal/doc/source/development/rfc/rfc76_ogrpythondrivers.rst

.. _rfc-76:

================================================================================
RFC 76: OGR Python drivers
================================================================================

============== ============================
Author:        Even Rouault
Contact:       even.rouault @ spatialys.com
Started:       2019-Nov-5
Last updated:  2019-Nov-5
Status:        Draft
Target:        GDAL 3.1
============== ============================

Summary
-------

This RFC adds the capability to write OGR/vector drivers in Python.

Motivation
----------

For some use cases that do not require lighting speed, or to deal with very
niche formats (possibly in house format), it might be faster and more efficient
to write a vector driver in Python rather than a GDAL C++ driver as currently required,
or an ad-hoc converter.

> This could benefit to QGIS which currently has no way to use providers in

Reviewer: Actually, you can create QGIS vector data provider in python: https://github.com/qgis/QGIS/blob/master/tests/src/python/test_provider_python.py

Author: good to know ! This probably didn't exist at the time I started this experiment, or I just didn't know :-)

    tools to use the OGR Python driver.

How does that work ?
--------------------

Driver registration
+++++++++++++++++++

The driver registration mechanism is extended to look for .py scripts in a
dedicated directory ({prefix}/lib/gdalplugins/python if hardcoded on compilation
for Unix builds, or the GDAL_PYTHON_DRIVER_PATH environment variable).
Those Python script must have in their first lines at least 2 global variables
set:

- ``DRIVER_NAME = "short_name"``
- ``DRIVER_SUPPORTED_API_VERSION = 1`` . Currently only 1 supported. If the
  interface changed in a backward incompatible way, we would increment internally
  the supported API version number. This item enables us to check if we are able
  to "safely" load a Python driver. If a Python driver would support several API
  versions (not clear if that's really possible at that point), it might use an
  array syntax to indicate that, like ``[1,2]``
- ``DRIVER_DCAP_VECTOR = "YES"``
- ``DRIVER_DMD_LONGNAME = "my super plugin"``

Optional metadata such as ``DRIVER_DMD_EXTENSIONS`` or
``DRIVER_DMD_HELPTOPIC`` can be defined (basically, any driver metadata key string prefixed by

> ``DRIVER_``

Reviewer 2: It may be (better clearer to reader and script maintainer, unambiguous for detection and loading) if those have `GDAL_` prefix.

Author: taken into account

Those will be parsed in a pure textual way, without invocation of the Python
interpreter, both for efficiency consideration and also because we want to
delay the research or launch of the Python interpreter as much as possible
(the typical use case if GDAL used by QGIS: we want to make sure that QGIS
has itself started Python, to reuse that Python interpreter)

From the short metadata, the driver registration code can instantiate GDALDriver
C++ objects. When the Identify() or Open() method is invoked on that object,
the C++ code will:

> * start/find the Python interpreter if not already done

Reviewer 2: What does the '"interpreter" mean?
A `PyThreadState` or `Py_NewInterpreter`-made sub-interpreter? 
An installed Python interpreter external to the Python built into GDAL?

IMO, this is a bit unclear.

Author: I've tried to make that a bit clearer

* load the .py file as a Python module in the Python interpreter if not already
* invoke the Python ``create_driver`` method, if not already done, that will
  instantiate a ``Driver`` object with a ``identify`` and ``open`` method.
* call the  ``identify`` and ``open`` method depending on the originated API call.

The ``open`` method will return a Python ``Dataset`` object with required and
optional methods that will be invoked by the corresponding GDAL API calls. And
likewise for the ``Layer`` object. See the example_.

Connection with the Python interpreter
++++++++++++++++++++++++++++++++++++++

The logic will be shared with the VRT Python fix functionality. It relies on
runtime loading of the Python library, rather than compile time linking. The
reason is that we do not know in advance with which Python version GDAL can
potentially be linked, and we do not want gdal.so/gdal.dll to be explicitly linked

> with a particular Python library.

Reviewer 2: This explains my earlier question in https://github.com/OSGeo/gdal/pull/1984/files#r342839730 (a footnote missing)

                    'datetimeField': '2017-04-26T12:34:56.789Z',
                    'geomField': 'POINT(2 49)'}

    class Dataset(GDALDataset):

        def __init__(self, filename):
            self.layers = [Layer()]
            pass

        # Optional, called on native object destruction
        def __del__(self):
            pass

        # Optional
        def metadata(self, domain):
            if domain is None:
                return {'foo': 'bar'}
            return None

        # Required, unless a layers attribute is set in __init__
        # def layer_count(self):
        #     return len(self.layers)

        # Required, unless a layers attribute is set in __init__
        # def layer(self, idx):
        #     return self.layers[idx]

    class Driver(GDALDriver):

        # Required
        def identify(self, filename, first_bytes, open_flags):
            return filename == 'MYPLUGIN:'

        # Required
        def open(self, filename, first_bytes, open_flags):
            if not self.identify(filename, first_bytes, open_flags):
                return None
            return Dataset(filename)

    # Required
    def create_driver():

>         return Driver()

Reviewer 2: Why this `create_driver` is required; what's the purpose?
Why not to use the `__init__` magic?

Since user's module is regularly imported
 https://github.com/rouault/gdal/blob/1bf795f25458dfb7e2a88e5d99b0e3647b82f2d4/gdal/gcore/gdalpythondriverloader.cpp#L2212-L2214

then why not look for classes that are `PyType_IsSubtype` of `Driver` class and construct calling the regular Python `__init__`?

```cpp
PyObject* poTypeDriver = ... ; // class item found with PyDict_Next for callable PyType_Check
                               // that is PyType_IsSubtype of Driver
PyTypeObject* poDriver = reinterpret_cast<PyTypeObject*>(poTypeDriver);
if (poDriver->tp_new(obj.get(), args, kwds) < 0) { ... }
if (poDriver->tp_init(obj.get(), args, kwds) < 0) { ... }

PyObject* poMethod = PyObject_GetAttrString(poDriver, "open" );
...

```

Author: sounds like a good idea. Adding a bit of implementation complexity though. Do you have an idea how to enumerate the classes of a module ? Can't see anything better than PyModule_GetDict() ( https://docs.python.org/3/c-api/module.html#c.PyModule_GetDict ) although they suggest using other methods, but not clear to me which one(s) would be appropriate

Reviewer 2: I usually do something along this lines (read the snippet as pseudo-code, not compile/run-tested, written from brain memory):

```cpp
PyObject *pKey(nullptr), *pValue(nullptr);
Py_ssize_t pos(0); // must be initialized to 0 prior call to PyDict_Next! 

PyObject* pModule = ...; // module with Driver, OGRLayer, etc.
PyObject* pDriverClass = PyObject_GetAttrString(pModule, "Driver");
PyTypeObject* pDriverType = reinterpret_cast<PyTypeObject*>(pDriverClass);

PyObject* pDict = PyModule_GetDict(pPluginModule, ...);
while (PyDict_Next(pDict, &pos, &pKey, &pValue))
{
    PyObject* pDictAttr = PyObject_GetAttr(pDict, pKey)
    if (PyType_Check(pDictAttr))
    {
        PyTypeObject* pType = reinterpret_cast<PyTypeObject*>(pDictAttr);
        if (PyType_IsSubtype(pDriverType, pType))
        {
            // User-defined Python class is viable candidate for Driver class
            // ... allocate/instantiate/call it...
        }
    }
}
```

> they suggest using other methods, but not clear to me which one(s) would be appropriate

I think, for `PyObject_*` they mean functions like `PyObject_GetAttrString`, but for `PyModule_*`, not sure.

Author: ok, I've managed to no longer require a create_driver() method. Using Python C API wasn't really possible as most of the PyType_xxxx methods are actually macros, which require having knowledge of Python structures, which isn't super ideal given that I don't include Python.h to be as much version independent as possible... So I went through the road of executing a Python snippet instead: https://github.com/rouault/gdal/commit/884ced57d8e720793e9f65f9eb059276d5db1c32#diff-42d4e6b0795db7beb7d99c0b9e51b18fR1190

Reviewer 2: >  PyType_xxxx methods are actually macros, which require having knowledge of Python structures (...) I don't include Python.h ...

Sorry, I ignored that fact indeed.

Limitations and scope
---------------------

- Vector and read-only for now. This could later be extended of course.

> - No connection between the Python code of the plugin and the OGR Python API 

Reviewer 2: Simply, the new one is "GDAL in Python"  and the existing one is "Python in GDAL" :-)

Reviewer 3: @mloskot or in other words, extending Python vs embedding Python: https://docs.python.org/3/extending/. I don't think it's a great idea for GDAL to promote both, but maybe that's just where GDAL is in the software universe, and the Python docs do cover the topic of extending embedded Python: https://docs.python.org/3/extending/embedding.html#extending-embedded-python.

Reviewer 2: Good point. Having written all three (extend, embed, extend embedded incl. extending with plain Python modules) and I don't think this proposal falls that into canonical embedded (and extend embedded) with Python.

Author: > extending Python vs embedding Python:

Discovering the taxonomy of those Python techniques :-) As far as I understand the docs, the techniques I use here are embedding Python (to load the user .py script, and the empty definitions of the OGRLayer, GDALDataset and GDALDriver classes) and extending python to define a _gdal_base module (internal detail) that has a layer_featureCount() method that is called by OGRLayer.feature_count() default implementation

> One can imagine scenarios where someone tries to use a Python driver in GDAL (embedded Python) and the Python driver uses the GDAL bindings (extending Python) and a VRT with Python functions (embedded Python again), and so on. The possibilities for bugs are endless.

I just tested that in https://gist.github.com/rouault/1dcc5a13583ba2ab419cf4a8f05b1ccc. It works fine from ogrinfo command line. Works fine also from a Python console doing "from osgeo import ogr; ds = ogr.Open('MYPLUGIN:')"  (so Python (console) calling C++ calling Python (OGR Python driver) calling C++ calling Python (embedded Python in VRT))  . I agree there might be some fragility though. Hence the limited scope to experimentation for now to see how it flyes with users.

> Or, it is to allow import x where x is not necessarily installed and available for embedded Python...

Yes, the Python drivers need to import gdal_base that has the base objects and some hidden magic to interact with the native code. This makes deployment easier rather than having a real .py + .so small module extending Python (like the class SWIG bindings do)

Reviewer 2: @rouault 
> As far as I understand the docs, the techniques I use here are embedding Python (to load the user .py script, and the empty definitions of the OGRLayer, GDALDataset and GDALDriver classes) and extending python to define a _gdal_base module (internal detail)  ...

In my opinion, the user-defined subclass of `Driver`, `OGRLayer` pre-defined classes are also an extension to the embedded Python.

The way you are proposing it is slightly more complex than canonical way of embedding Python:
it is not linking  Python library at build time in order to embed the Python interpreter and API, but actually embedding Python at run-time. I'd call it an extremely delayed or lazy embedding :)

I would not be surprised if some readers of the RFC and users (especially not experienced w/ embedding) of the machinery are a bit confused asking on the list how it works, because it is not a typical way of embedding.

Reviewer 4: > The possibilities for bugs are endless.

PDAL both extends and embeds Python. This statement is a fact. It is also really useful to do both things sometimes.

Author: few words added about that

Testing
-------

The gdalautotest suite will be extended with a simple test Python driver, along
the lines of the above example.

Previous discussions
--------------------

This topic has been discussed in the past in :

- https://lists.osgeo.org/pipermail/gdal-dev/2017-April/thread.html#46526
- https://lists.osgeo.org/pipermail/gdal-dev/2018-November/thread.html#49294

Implementation
--------------

A preliminary implementation is available at in
https://github.com/rouault/gdal/tree/pythondrivers

https://github.com/OSGeo/gdal/compare/master...rouault:pythondrivers

It will be modified to have common code between it and the existing VRT Python
pixel function mechanism regarding how the Python interpreter is detected/loaded.

Topics for discussion
---------------------

* Currently that code is unconditionnally compiled in, similarly to the GDAL
  VRT Python plugin function. Is there a need to have a compile-time switch
  to disable this ?

* Should we use __iter__() / __next__() methods in the Layer object to iterate
  over features ? This is the canonical way in Python to define an iterator.
  However, this code is to be called mainly by the GDAL C++ code, so Python'icity

>   is not critical. Another subtlety is that OGR layers can have spatial and

Reviewer 2: I can be someone will complain why the API is not Pythonic :-)

Author: I opted for a Python iterator finally

-------

This RFC adds the capability to write OGR/vector drivers in Python.

Motivation
----------

For some use cases that do not require lighting speed, or to deal with very
niche formats (possibly in house format), it might be faster and more efficient
to write a vector driver in Python rather than a GDAL C++ driver as currently required,
or an ad-hoc converter.

.. note::

    QGIS has now a way to create Python-based providers such as
    in https://github.com/qgis/QGIS/blob/master/tests/src/python/provider_python.py
    Having a way to do in GDAL itself also allows the rest of GDAL/OGR based
    tools to use the OGR Python driver.

How does that work ?
--------------------

Driver registration
+++++++++++++++++++

The driver registration mechanism is extended to look for .py scripts in a
dedicated directory:

* the directory pointed by the ``GDAL_PYTHON_DRIVER_PATH`` configuration option
  (there may be several paths separated by `:` on Unix or `;` on Windows)
* if not defined, the directory pointed by the ``GDAL_DRIVER_PATH`` configuration
  option.
* if not defined, in the directory (hardcoded at compilation time on Unix builds)
  where native plugins are located.

> Those Python script must have in their first lines at least 2 global variables

Reviewer 3: ```suggestion
Those Python script must set in their first lines at least 2 directives
```

> set:

Reviewer 3: ```suggestion
```

- ``# gdal: DRIVER_NAME = "short_name"``
- ``# gdal: DRIVER_SUPPORTED_API_VERSION = 1`` . Currently only 1 supported. If the
  interface changed in a backward incompatible way, we would increment internally
  the supported API version number. This item enables us to check if we are able
  to "safely" load a Python driver. If a Python driver would support several API
  versions (not clear if that's really possible at that point), it might use an
  array syntax to indicate that, like ``[1,2]``
- ``# gdal: DRIVER_DCAP_VECTOR = "YES"``
- ``# gdal: DRIVER_DMD_LONGNAME = "my super plugin"``

Optional metadata such as ``# gdal: DRIVER_DMD_EXTENSIONS`` or
``# gdal: DRIVER_DMD_HELPTOPIC`` can be defined (basically, any driver metadata key string prefixed by
``# gdal: DRIVER_``

> Those will be parsed in a pure textual way, without invocation of the Python

Reviewer 3: ```suggestion
These directives will be parsed in a pure textual way, without invocation of the Python
```

They are minimal. The GDALAllRegister() method has an added call to
GDALDriverManager::AutoLoadPythonDrivers() that implements the above mentioned
logic. The GDALDriver class has been extended to support a new function
pointer, IdentifyEx(), which is used by the C++ shim that loads the Python code.

.. code-block:: c++

    int                 (*pfnIdentifyEx)( GDALDriver*, GDALOpenInfo * );

This extended IdentifyEx() function pointer, which adds the GDALDriver* argument,
is used in priority by GDALIdentify() and GDALOpen() methods. The need for that
is purely boring. For normal C++ drivers, there is no need to pass the driver,
as there is a one-to-one correspondence between a driver and the function that
implements the driver. But for the Python driver, there is a single C++ method
that does the interface with the Python Identify() method of several Python drivers,
hence the need of a GDALDriver* argument to forward the call to the appropriate
driver.

.. _example:

Example of such a driver
------------------------

Note that the prefixing by the driver name in the connection string is absolutely
not a requirement, but something specific to this particular driver which is a

> bit artifical. The CityJSON driver mentionned below does not need it.

Reviewer 3: ```suggestion
bit artifical. The CityJSON driver mentioned below does not need it.
```

.. code-block:: python

    #!/usr/bin/env python
    # -*- coding: utf-8 -*-
    # This code is in the public domain, so as to serve as a template for
    # real-world plugins.
    # or, at the choice of the licensee,
    # Copyright 2019 Even Rouault
    # SPDX-License-Identifier: MIT

    # The gdal_python_driver module is defined by the GDAL library at runtime
    from gdal_python_driver import BaseDriver, BaseDataset, BaseLayer

    # Metadata parsed by GDAL C++ code at driver pre-loading, starting with '# gdal: '
    # Required and with that exact syntax since it is parsed by non-Python
    # aware code. So just literal values, no expressions, etc.
    # gdal: DRIVER_NAME = "DUMMY"
    # API version(s) supported. Must include 1 currently
    # gdal: DRIVER_SUPPORTED_API_VERSION = [1]
    # gdal: DRIVER_DCAP_VECTOR = "YES"
    # gdal: DRIVER_DMD_LONGNAME = "my super plugin"

    # Optional driver metadata items.
    # # gdal: DRIVER_DMD_EXTENSIONS = "ext1 est2"

>     # # gdal: DRIVER_DMD_HELPTOPIC = "http://example.com/my_help.html"

Reviewer 3: Recommend putting these directives above the Python import statement.

Author: done
